### PR TITLE
cmake/toolchain: Support LLVM-style profiling/coverage for native_posix

### PR DESCRIPTION
**Pull request by andyross**
Downstream POSIX build consumers (specifcially oss-fuzz) are set up to do analysis on LLVM coverage data.  Provide a simple kconfig to turn that on for them.

In theory this is available for any LLVM/clang setup, though in practice only host builds are working.  The infrastructure and docs to extract the profiling data from embedded target builds are... thin.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1976,6 +1976,11 @@ if(CONFIG_MINIMAL_LIBCPP)
   zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler-cpp,nostdincxx>>)
 endif()
 
+if(CONFIG_LLVM_PROFILE)
+  zephyr_compile_options(-fprofile-instr-generate -fcoverage-mapping)
+  zephyr_link_libraries(-fprofile-instr-generate -fcoverage-mapping)
```
Comment (andyross) [inline CMakeLists.txt:1981]: One of the cmake experts will need to tell me if this is the right spot for this or not.  It seems weird to have all the compiler arguments here at the top level, but putting it into the cmake/toolchain/ tree didn't work because the default target there is "zephyr_interface" (or something to that effect).

**Comment by andyross** [comment]
FWIW: looks like the compilance failure is a false positive because the new kconfig (which exists only when ZEPHYR_TOOLCHAIN_VARIANT=llvm) isn't included in the build?  Or should I move it somewhere else?

File: CMakeLists.txt (hunk 1 above)
Comment (tejlmand) [reply CMakeLists.txt:1981]: can you elaborate, what did not work / how did it not work ?

zephyr_compile_options and zephyr_link_libraries both works on the Zephyr interface, so there should not be any difference there.

File: cmake/toolchain/llvm/Kconfig [hunk 2]
```diff
@@ -17,3 +17,15 @@ config LLVM_USE_LLD
 	  Use LLVM built-in lld linker with llvm/clang.
 
 endchoice
+
+config LLVM_PROFILE
```
Comment (tejlmand) [inline cmake/toolchain/llvm/Kconfig:21]: Should this really be named `LLVM_PROFILE` ?

I acknowledge it's different than our current `COVERAGE` feature, but it is still desirable that we try to keep names more general so that the Kconfig name can cover more that just a single toolchain, and so far, the `-fprofile-instr-generate -fcoverage-mapping` is still coverage, just a different variant / output.

I don't have a proposal atm, but would like to prevent situations where each toolchain has their own setting names for features that should be general (although their exact implementation / flags may differ).

**Comment by cfriedt** [comment]
Nifty ;-) I'll give it an ack whenever Thorsten is ok with it.

**Comment by andyross** [comment]
Rebase, respond to comments that I'd missed.

If there's genuine resistance, I can certainly make this work with the existing CONFIG_COVERAGE, by e.g. detecting that we're using it with a LLVM toolchain and turning on/off features as needed.  But for sure that's going to be more code, for what is really a pretty simple thing.  All I want here is a way for downstream apps (oss-fuzz integration in particular) to turn it on in a convenient way without mucking with global cmake settings.

Regarding that: when I put exactly this code block into the cmake/toolchain/llvm/target.cmake file, I get an error back about wrong target or something.  I assumed this was just a known issue with the way the files are organized, because it turns out that basically all global compiler switches are here, in the top level file I'm modifying.  No idea, basically.  But if it's a bug it's one we've decided to work around by putting this here.

**Comment by andyross** [comment]
Quick ping for re-review?  Really this wasn't supposed to be a complicated thing.  It's a toolchain-specific flag with an obvious way to turn it on.  For sure we can look at doing a long-term portable coverage implementation later, but all I want here is for oss-fuzz to generate coverage binaries without having to patch the cmake code in the scripting downstream.

File: cmake/toolchain/llvm/Kconfig (hunk 2 above)
Comment (cfriedt) [reply cmake/toolchain/llvm/Kconfig:21]: Is there no way to determine if the toolchain is LLVM or gcc-compatible or otherwise?

It seems like this would be trivial to do (at a low level) if that were the case. I guess getting  things in similar reporting formats is more of a twister problem

**Comment by andyross** [comment]
Ping again.  To repeat, this is holding up fuzzing work on SOF.  I'm fine with dithering about design to an extent, but it's not really clear what the desired implementation even is.  Can we at least merge this now and open a bug for rework later?

**Comment by nordicjm** [comment]
Torsten is on holiday until the end of the month

File: cmake/toolchain/llvm/Kconfig [hunk 3]
```diff
@@ -17,3 +17,15 @@ config LLVM_USE_LLD
...
+	bool "Enable profiling output"
+	depends on ARCH_POSIX
+	help
+	  Builds the (posix) app with profiling and instruction
```
Comment (aescolar) [inline cmake/toolchain/llvm/Kconfig:25]: A minor thing, The `Builds the (posix) app with..` can be a bit confusing.
It is everything we build (not just the app).
Maybe better to just say `Builds with profiling and instruction`
(You are mention below it only works for ARCH_POSIX)

File: CMakeLists.txt (hunk 1 above)
Comment (aescolar) [reply CMakeLists.txt:2008]: (If you do what I commented in the other comment, redefining $<TARGET_PROPERTY:{compiler/linker},coverage>, you should not need to do this change.)

But if you want to go this way, I'd expect you could put it around https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/posix/CMakeLists.txt#L108  without problems(?)

And if you do so, please add also the equivalent 
```
target_compile_options(native_simulator INTERFACE ..)
target_link_options(native_simulator INTERFACE ..)
```
so it also works for native_sim

File: CMakeLists.txt [hunk 4]
```diff
@@ -2003,6 +2003,11 @@ if(CONFIG_MINIMAL_LIBCPP)
   zephyr_compile_options($<$<COMPILE_LANGUAGE:CXX>:$<TARGET_PROPERTY:compiler-cpp,nostdincxx>>)
 endif()
 
+if(CONFIG_LLVM_PROFILE)
+  zephyr_compile_options(-fprofile-instr-generate -fcoverage-mapping)
```
Comment (aescolar) [inline CMakeLists.txt:2007]: This should probably be commented by @tejlmand (meaning, if he proposes something else, ignore what I say here), but would it not be nicer if instead of doing this, 
We'd have the definition of these two
`$<TARGET_PROPERTY:compiler,coverage>` and `$<TARGET_PROPERTY:linker,coverage>`
be based on the kconfig choice?

Note: `compiler,coverage` is defined in:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/gcc/compiler_flags.cmake#L156-L157
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/compiler/clang/compiler_flags.cmake#L17-L18
And `linker,coverage` in:
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/linker/ld/clang/linker_flags.cmake
https://github.com/zephyrproject-rtos/zephyr/blob/main/cmake/linker/ld/gcc/linker_flags.cmake#L7

File: cmake/toolchain/llvm/Kconfig (hunk 2 above)
Comment (aescolar) [reply cmake/toolchain/llvm/Kconfig:21]: What about a choice dependent on COVERAGE, of COVERAGE_NATIVE_SOURCE or COVERAGE_NATIVE_GCOV or COVERAGE_GCOV.
where:
* COVERAGE_NATIVE_GCOV would be the default with NATIVE_BUILD (what we have today)
* COVERAGE_NATIVE_SOURCE is this new LLVM source coverage mode
* COVERAGE_GCOV is the current COVERAGE_GCOV (embedded gcov data generation).

And with COVERAGE_NATIVE_SOURCE being dependent on LLVM (`depends on "${ZEPHYR_TOOLCHAIN_VARIANT}" = "llvm"`) and NATIVE_BUILD.
And with COVERAGE_NATIVE_GCOV  being dependent NATIVE_BUILD.

**Review by aescolar — COMMENTED** [review]
A proposal

**Comment by andyross** [comment]
Weekly ping.  Can we get a +1 or at least a specific instruction as to the right way to do this?  :)

I'll say again, "adding a compiler flag for a specific downstream use case" should **really** not require a months-long review cycle...

**Review by teburd — APPROVED** [review]

File: CMakeLists.txt (hunk 4 above)
Comment (tejlmand) [reply CMakeLists.txt:2007]: I think it makes a lot of sense to use the `coverage` property for this.

That will keep the code in common CMakeLists.txt generic and thus allow other toolchains decide / implement the flags to use which provides the desired functionality.

**Comment by tejlmand** [comment]
> Regarding that: when I put exactly this code block into the cmake/toolchain/llvm/target.cmake file, I get an error back about wrong target or something. I assumed this was just a known issue with the way the files are organized, because it turns out that basically all global compiler switches are here, in the top level file I'm modifying. 

Ok, that explains / clarifies what was not working.

Generally we try to minimize specific toolchain (compiler/linker) handling in the common Zephyr CMakeLists.txt code, in order to avoid conditional code like:
```
if(gcc)
...
elseif(llvm)
...
elseif(arcmwdt)
.... and so on
```

instead the toolchain (compiler / linker) sets specific properties for compiler and/or linker which can then be used as:
```
zephyr_compile_options($<TARGET_PROPERTY:compiler,<prop-name>>)
zephyr_link_libraries($<TARGET_PROPERTY:linker,<prop-name>>)
```
it also allows us to have generic Kconfig settings names as each toolchain implementation knows what flags to use to achieve the desired behavior intended with the Kconfig setting, for example what `CONFIG_COVERAGE` or `CONFIG_PROFILE` means.

It can be done like @aescolar proposes here: https://github.com/zephyrproject-rtos/zephyr/pull/59632#discussion_r1269203051

**Comment by andyross** [comment]
> it also allows us to have generic Kconfig settings names as each toolchain implementation knows what flags to use to achieve the desired behavior intended with the Kconfig setting, for example what CONFIG_COVERAGE or CONFIG_PROFILE means.

But... again, I'm not implementing CONFIG_COVERAGE as a feature. It would be good if I was!  But I'm not.  I don't have time and it's not something I'm interested in or good at.  You're demanding abstraction for something that can't be abstracted.

I need to enable coverage reporting using downstream tooling (oss-fuzz), which is supposed to be as simple as "just enable this compiler flag".  And... we're seriously refusing to support that kind of use case?

Can you **please** give me a specific set of instructions for code you'll accept that don't involve trying to duplicate CONFIG_COVERAGE's semantics?  If not I guess I have to close this and either patch zephyr in the oss-fuzz scripting or muck with EXTRA_CFLAGS or something.  I remain dumbfounded that something this simple is being rejected and has no path to merge after almost two months of review.

**Comment by andyross** [comment]
Wow, OK, I'm going to try one last time:

That gives me a CONFIG_COVERAGE=y build.  Zephyr **does not build the same** with CONFIG_COVERAGE=y.  What you get is a build tailored to the particular needs of Zephyr's own coverage reporting.  It requires board support in a few places.  It has whitebox spots where we seem to be trying to exercise cases that can't be reached.  It has bits where we're stubbing things out that would otherwise be false positives, etc...  Just grep for COVERAGE in the tree and you can see all these spots.  Stack sizes grow to accomodate the gcc profiling data, etc...

I don't want any of that.  I want a **normal** Zephyr build that **happens to have** two command line options passed to the compiler for the benefit of downstream tooling.  And, OK, sure, those flags happen to be ones that are also used for the equivalent of coverage testing in the LLVM world.  But that's fundamentally just a coincidence!

Now, that all being said: my personal read is that none of the deltas are going to meaningfully affect anything in a SOF fuzzing build.  So... yeah, sure.  I guess we could merge this and it would work.  But it would "work" only until someone comes along and changes the way COVERAGE semantics work, which is a fragile dependency SOF fuzzing never needed and adoesn't want.

Basically: this is bad abstraction.  It's the wrong design for this case.  I want control over two command line options in a supported toolchain, I should be able to have control over two command line options in a supported toolchain without conforming to the requirements of a subsystem that I'm not implementing.

I'm booked today, but I'll get back to the fuzzing by tomorrow.  If you guys won't bend we can merge the "overload COVERAGE incorrectly" version. :)

**Comment by aescolar** [comment]
@andyross Should be then close this one? (now that https://github.com/zephyrproject-rtos/zephyr/pull/61892 is merged)

**Review by cfriedt — APPROVED** [review]
I'd like to see this get in. Can we get a rebase?

**Comment by aescolar** [comment]
@cfriedt isn't #61892 providing you already what you need?

**Comment by cfriedt** [comment]
> @cfriedt isn't #61892 providing you already what you need?

Ah! Yes. I guess we can close this one rhen
